include/openssl/core.h

```diff
@@ -136,6 +136,26 @@ struct ossl_param_st {
 # define OSSL_PARAM_OCTET_STRING_PTR                    \
     (OSSL_PARAM_OCTET_STRING|OSSL_PARAM_POINTER_FLAG)
 
+/*-
+ * Provider entry point
+ * --------------------
+ *
+ * This function is expected to be present in any dynamically loadable
+ * provider module.  By definition, if this function doesn't exist in a
+ * module, that module is not an OpenSSL provider module.
+ */
+/*-
+ * |provider|   pointer to opaque type OSSL_PROVIDER.  This can be used
+ *              together with some functions passed via |in| to query data.
+ * |in|         is the array of functions that the Core passes to the provider.
+ * |out|        will be the array of base functions that the provider passes
+ *              back to the Core.
+ */
+typedef int (ossl_provider_init_fn)(const OSSL_PROVIDER *provider,
+                                    const OSSL_DISPATCH *in,
+                                    const OSSL_DISPATCH **out);
+extern ossl_provider_init_fn OSSL_provider_init;
+
```

Comment:
This seems incorrect naming - in that it should probably be OSSL_PROVIDER_provider_init as the entry point ...

Reply:
Why?  It's not a function for handling the OSSL_PROVIDER object.

Reply:
This is the entry point of a provider module, and it is the name we went for for a long time in the [design document](https://www.openssl.org/docs/OpenSSL300Design.html).  It *isn't* a method for the `OSSL_PROVIDER` type per se, so that's not a meaningful reason for such a name change.

(`OSSL_PROVIDER` is a libcrypto side abstraction of a provider, be it built in or in a module, and it does have a set of non-public methods)

Reply:
I agree with @levitte. The `FOO_blah(FOO *, ...)` naming convention is really for where `FOO_blah` could be considered a method of the object `FOO`. But that isn't the case here. So in fact in my mind it would be misleading to call it `OSSL_PROVIDER_provider_init` - that gives the wrong impression to the reader of the code.